### PR TITLE
Add constexpr to Point and remove redundant operator

### PR DESCRIPTION
**Pull request by RichardYCJ**
Related to #16243.

### CODE REVIEW

**Review by masterleinad — COMMENTED** [review]
Please add a test demonstrating that all these functions can be used in a `constexpr` context.

File: include/deal.II/base/point.h [hunk 1]
```diff
@@ -201,96 +202,13 @@ class Point : public Tensor<1, dim, Number>
...
    * @{
    */
-
-  /**
-   * Add an offset given as Tensor<1,dim,Number> to a point.
-   *
-   * @note This function can also be used in @ref GlossDevice "device" code.
-   */
-  DEAL_II_HOST_DEVICE Point<dim, Number>
-                      operator+(const Tensor<1, dim, Number> &) const;
```
Comment (bangerth) [inline include/deal.II/base/point.h:218]: I don't think you can just remove this function. If you do, the compiler will choose the `operator+` in the base class, which returns a `Tensor<1,dim>`, not a `Point<dim>`. Or do I miss something?

File: include/deal.II/base/point.h [hunk 2]
```diff
@@ -646,6 +463,57 @@ inline void Point<dim, Number>::serialize(Archive &ar, const unsigned int)
...
+ */
+template <int dim, typename Number>
+DEAL_II_CXX20_REQUIRES(dim >= 0)
+constexpr DEAL_II_HOST_DEVICE Point<dim, Number>
+operator+(const Point<dim, Number> &p, const Tensor<1, dim, Number> &offset)
+{
+  Point<dim, Number> tmp{p};
+  tmp += offset;
+  return tmp;
+}
```
Comment (bangerth) [inline include/deal.II/base/point.h:479]: Ah, I see -- you are replacing the member function by a global function. It's not clear to me why this is advantageous, though. What is the rationale?

**Review by bangerth — COMMENTED** [review]
The `constexpr` changes by themselves are ok, but as already mentioned, it would be nice to have tests.

The operator changes are not clear to me. If you wanted to remove these from the patch for the moment, I think we can make progress faster -- one issue per pull request :-)

**Comment by RichardYCJ** [comment]
> Please add a test demonstrating that all these functions can be used in a `constexpr` context.

I'm doing it, but I have to learn how to add test first :rofl:

File: include/deal.II/base/point.h (hunk 2 above)
Comment (RichardYCJ) [reply include/deal.II/base/point.h:479]: Ah, I remember some posts, say [this stackoverflow post](https://stackoverflow.com/questions/4622330/operator-overloading-member-function-vs-non-member-function), suggest that it is better to write the binary operator as a non-member function considering the invoke order.

Here it does not matter too much, since we have Tensor operator in the bottom anyway. But it is a better convention from my perspective.

Comment (bangerth) [reply include/deal.II/base/point.h:479]: I think this is perhaps debatable. The stackoverflow post you point out is of course correct if you want an operator where the *left* operand is a non-class type, but that isn't the case here.

There is, however, a good reason to leave the code as it is: The documentation of this function will clearly show up as part of the documentation of the class. With global functions, doxygen will create a page for the function, but it may not be linked to the documentation of the `Point` class.

Comment (RichardYCJ) [reply include/deal.II/base/point.h:479]: You are right. There's no difference in this situation but since `Tensor` also leave its `operator+-*/` as a global function, maybe we should also do it with `Point`?

We could use `@relates` to hint doxygen to link `operator+-` and it will only appear on that page just like other functions. What do you think?

Comment (bangerth) [reply include/deal.II/base/point.h:479]: Yes, if you can make it look essentially like `Tensor`, then I'm in favor :-)

**Comment by RichardYCJ** [comment]
I add several static_assert in `point_01` test, not sure whether its suitable. First trial on using test 🥲

**Comment by RichardYCJ** [comment]
It seems the error message in Linux serial build is not related to this PR?

**Comment by masterleinad** [comment]
```
In file included from /__w/dealii/dealii/include/deal.II/grid/reference_cell.h:22,
                 from /__w/dealii/dealii/include/deal.II/fe/fe_data.h:23,
                 from /__w/dealii/dealii/include/deal.II/fe/fe.h:23,
                 from /__w/dealii/dealii/include/deal.II/fe/fe_poly.h:25,
                 from /__w/dealii/dealii/include/deal.II/fe/fe_pyramid_p.h:23,
                 from /__w/dealii/dealii/source/grid/reference_cell.cc:21:
In static member function ‘static unsigned int dealii::GeometryInfo<dim>::child_cell_on_face(const dealii::RefinementCase<dim>&, unsigned int, unsigned int, bool, bool, bool, const dealii::RefinementCase<(dim - 1)>&) [with int dim = 2]’,
    inlined from ‘static unsigned int dealii::GeometryInfo<dim>::face_to_cell_vertices(unsigned int, unsigned int, bool, bool, bool) [with int dim = 2]’ at /__w/dealii/dealii/include/deal.II/base/geometry_info.h:4763:28,
    inlined from ‘unsigned int dealii::ReferenceCell::face_to_cell_vertices(unsigned int, unsigned int, unsigned char) const’ at /__w/dealii/dealii/include/deal.II/grid/reference_cell.h:1899:56,
    inlined from ‘dealii::Point<dim> dealii::ReferenceCell::closest_point(const dealii::Point<dim>&) const [with int dim = 3]’ at /__w/dealii/dealii/source/grid/reference_cell.cc:1020:46:
/__w/dealii/dealii/include/deal.II/base/geometry_info.h:4444:57: error: ‘subcells[0][2][<unknown>][2]’ may be used uninitialized [-Werror=maybe-uninitialized]
 4444 |   return subcells[face_flip][ref_case - 1][face][subface];
      |                                                         ^
/__w/dealii/dealii/include/deal.II/base/geometry_info.h: In member function ‘dealii::Point<dim> dealii::ReferenceCell::closest_point(const dealii::Point<dim>&) const [with int dim = 3]’:
/__w/dealii/dealii/include/deal.II/base/geometry_info.h:4429:5: note: ‘subcells’ declared here
 4429 |     subcells[2][RefinementCase<2>::isotropic_refinement][faces_per_cell]
      |     ^~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [source/grid/CMakeFiles/object_grid_release.dir/build.make:160: source/grid/CMakeFiles/object_grid_release.dir/reference_cell.cc.o] Error 1
make[2]: Leaving directory '/__w/dealii/dealii/build'
make[1]: *** [CMakeFiles/Makefile2:1981: source/grid/CMakeFiles/object_grid_release.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by masterleinad** [comment]
> It seems the error message in Linux serial build is not related to this PR?

We don't see it in other pull requests so the best bet is that it's related to this pull request.

File: include/deal.II/base/geometry_info.h [hunk 3]
```diff
@@ -4425,7 +4426,7 @@ GeometryInfo<2>::child_cell_on_face(const RefinementCase<2> &ref_case,
   // this cell adjacent to the subface of a possibly
   // refined neighbor. this simplifies setting neighbor
   // information in execute_refinement.
-  const unsigned int
+  constexpr unsigned int
     subcells[2][RefinementCase<2>::isotropic_refinement][faces_per_cell]
             [max_children_per_face] = {
```
Comment (bangerth) [inline include/deal.II/base/geometry_info.h:4431]: This is the change that causes the compiler error. Just remove the changes to `geometry_info.h` from this patch -- they are unrelated to what you want to achieve here anyway, so they should go into separate pull requests.

**Comment by RichardYCJ** [comment]
```
diff --git a/include/deal.II/base/geometry_info.h b/include/deal.II/base/geometry_info.h
index cf53e3351..[7](https://github.com/dealii/dealii/pull/16251/checks#step:5:8)b4df32ee 100644
--- a/include/deal.II/base/geometry_info.h
+++ b/include/deal.II/base/geometry_info.h
@@ -2923,7 +2923,7 @@ RefinementCase<dim>::serialize(Archive &ar, const unsigned int)
   // serialization can't deal with bitfields, so copy from/to a full sized
   // std::uint[8](https://github.com/dealii/dealii/pull/16251/checks#step:5:9)_t
   std::uint8_t uchar_value = value;
-  ar & uchar_value;
+  ar          &uchar_value;
   value = uchar_value;
 }
```

I use the `./contrib/utilities/download_clang_format` as the document formatter, why would the indentation still fail? Besides the format that CI suggests is incorrect since its `operator&` not a reference.

Should I revert this change or do we need to update the CI environment?

**Review by bangerth — COMMENTED** [review]
I looked through the patch again, and here are my current sticking points:

* You also remove the unary-minus operator from the `Point` class, but I do not think that you add a global version of it back. 
* I don't think you implement `operator*` and `operator/` right now.
* For the test, the `static_assert` is a nice idea! We typically create new tests instead of adding to old tests. In your case, create a new test -- say `point_constexpr.cc` -- that has the same framework as the `_01` test you modify but only has the `static_assert` statements and at the end outputs "OK" (see `point_02.cc` for example). Then you need a `point_constexpr.output` file that also only contains the "OK" text (again see `point_02.output`).
* Your branch has several merge commits. Don't do this. It confuses the history. Just start development at one point on `master` and stick with it. If you need to update the point on which your development is based, use `git rebase master` instead of `git merge master`.

**Comment by bangerth** [comment]
On 11/8/23 07:23, Chengjiang Yin wrote:
> ||
> 
> I use the |./contrib/utilities/download_clang_format| as the document 
> formatter,

This only *downloads* the formatter. You need to run 
`./contrib/utilities/indent` to actually do the indenting.

> why would the indentation still fail? Besides the format that CI 
> suggests is incorrect since its |operator&| not a reference.

Yes. It's a quirk of the tool that it does not recognize this correctly. But 
in all of the other places, it is still far preferable to manually indenting.

File: include/deal.II/base/geometry_info.h (hunk 3 above)
Comment (RichardYCJ) [reply include/deal.II/base/geometry_info.h:4431]: Oh, this is what I add to avoid the test error. I have to check it by eyes since my compiler didn't complain anything. Can I just leave this if it pass the test? I'm afraid any commit would break this again 😢

**Comment by RichardYCJ** [comment]
> You also remove the unary-minus operator from the Point class, but I do not think that you add a global version of it back.

That is intended. Since unary operator- is equal to `(-1)*`. If put it stricter in mathematics, Point (a location in space) itself has no definition of scalar multiplication but **can be defined on a Tensor from the origin to the point**. Actually in the code they don't make any difference since Tensor and Point can convert to each implicitly (I also remvoe explicit in ctor of `Point<dim>` from a `Tensor<1,dim>`)

> This only *downloads* the formatter. You need to run `./contrib/utilities/indent` to actually do the indenting.

I use `clang-format` extension in VSCode to format on save, I've already set the formatter to the downloaded one but somehow it still gives out different format.

File: include/deal.II/base/geometry_info.h (hunk 3 above)
Comment (bangerth) [reply include/deal.II/base/geometry_info.h:4431]: Well, for some reason it fails the Continuous Integration checks, and so we cannot merge it. It is also unrelated. If you want to make this change, please make it in a separate pull request and take it out of this one.

**Comment by bangerth** [comment]
>     You also remove the unary-minus operator from the Point class, but I
>     do not think that you add a global version of it back.
> 
> That is intended. Since unary operator- is equal to |(-1)*|. If put it 
> stricter in mathematics, Point (a location in space) itself has no 
> definition of scalar multiplication but *can be defined on a Tensor from 
> the origin to the point*. Actually in the code they don't make any 
> difference since Tensor and Point can convert to each implicitly (I also 
> remvoe explicit in ctor of |Point<dim>| from a |Tensor<1,dim>|)

That too is something we can't merge. We made the constructor explicit 
for a reason: A tensor is not a point, and we do not want to allow the 
implicit conversion.

Whether the addition of unary operator-() is a good idea or not is in 
some sense not an important question today. The decision that we wanted 
to support this operation was made more than 20 years ago, and we want 
to remain backward-compatible where possible. I don't see a particularly 
good reason why we should break backward compatibility here.


>     This only /downloads/ the formatter. You need to run
>     |./contrib/utilities/indent| to actually do the indenting.
> 
> I use |clang-format| extension in VSCode to format on save, I've already 
> set the formatter to the downloaded one but somehow it still gives out 
> different format.

I don't know about the format-on-save thing. I would suggest to just do 
it by hand on the command line before committing.

**Comment by RichardYCJ** [comment]
Sure, I will close this PR and make a new one only adds constexpr to Point with separate unit test.

**Comment by RichardYCJ** [comment]
Sure, I will close this PR and make a new one only add constexpr to Point with separate unit test.
